src/identity/customca.rs

```diff
@@ -0,0 +1,141 @@
+    }
+}
+
+#[async_trait]
+impl CertificateProvider for CustomSigner {
+    #[instrument(skip_all)]
+    async fn fetch_certificate(&self, id: &Identity) -> Result<tls::Certs, Error> {
+        let client = self.client.clone();
+        let csr: Api<CertificateSigningRequest> = Api::all(client.clone());
+        let pp = PostParams::default();
+        let cs = tls::CsrOptions {
+            san: id.to_string(),
+        }
+        .generate()?;
+        let cert: Vec<u8> = cs.csr;
+        let pkey = cs.pkey;
+        let csr_name = generate_csr_name();
+        let stored_csr_name = csr_name.clone();
+        let signer_name = &self.signer;
+        let mut usage = Vec::new();
+        usage.push("client auth".to_string());
+
+        let obj_metadata = k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta{
+            name: Some(csr_name),
+            ..Default::default()
+        };
+        let csr_spec = CertificateSigningRequestSpec{
+            expiration_seconds:Some(86400),
+            request:k8s_openapi::ByteString(cert),
+            signer_name:signer_name.as_ref().unwrap().to_string(),
+            usages:Some(usage),
+            ..Default::default()
+        };
+
+        let csr_req = CertificateSigningRequest {
+            metadata: obj_metadata,
+            spec: csr_spec,
+            ..Default::default()
+        };
+
+        // creating CSR
+        match csr.create(&pp, &csr_req).await {
+            Ok(o) => {
+                let name = o.metadata.name.unwrap_or_default();
+                assert_eq!(csr_req.metadata.name.unwrap_or_default(), name);
+            }
+            Err(kube::Error::Api(ae)) => assert_eq!(ae.code, 409),
+            Err(_) => return Err(Error::EmptyResponse(id.clone()))?
+        };
+
+        let final_certs = check_signed_cert(&csr, &stored_csr_name).await;
+        match final_certs {
+            Ok(cert_chain) => {
+                let cert_chain_vec = cert_chain.0;
+                let final_cert_chain = vec![str::from_utf8(&cert_chain_vec).unwrap()];
+                let leaf = final_cert_chain[0].as_bytes();
+                let final_chain = if final_cert_chain.len() > 1 {
+                    final_cert_chain[1..].iter().map(|final_cert_chain| final_cert_chain.as_bytes()).collect()
+                } else {
+                    warn!("no chain certs for: {}", id);
+                    vec![]
+                };
+                let res_certs = tls::cert_from(&pkey, leaf, final_chain);
+                res_certs
+                .verify_san(id)
+                .map_err(|_| Error::SanError(id.clone()))?;
+                return Ok(res_certs);
+            },
+            Err(_) => return Err(Error::EmptyResponse(id.clone()))?,
+        }
+    }
+}
+
+fn generate_csr_name() -> String{
+    let mut rng = thread_rng();
+    let rand_number: u16 = rng.gen();
+    let rand_string = rand_number.to_string();
+    let base_string = "csr-ztunnel-";
```

Comment:
would it make sense to include the service-account name in the csr request instead of keeping it completely random? We can update this in a future change if needed

Reply:
sure, will add this Identity as well - https://github.com/istio/ztunnel/blob/master/src/identity/manager.rs#L35

Reply:
I don't think we need to generate a random string manually, you can set generateName in API Create request instead of Name and have k8s do it for you. Thati s how Pod works, for example

Reply:
we need the resource name while watching the CSR, not sure if we let K8s to create the random name how will we watch the CSR ?

Reply:
the Create call returns the resulting name

Reply:
got it, thanks